File: src/main/java/de/blazemcworld/fireflow/code/node/impl/action/player/SendMessageNode.java
package de.blazemcworld.fireflow.code.node.impl.action.player;

import de.blazemcworld.fireflow.code.node.Node;
import de.blazemcworld.fireflow.code.type.PlayerType;
import de.blazemcworld.fireflow.code.type.SignalType;
import de.blazemcworld.fireflow.code.type.TextType;
import de.blazemcworld.fireflow.code.value.PlayerValue;
import net.kyori.adventure.text.Component;
import net.minestom.server.item.Material;

public class SendMessageNode extends Node {

    public SendMessageNode() {
        super("send_message", Material.PAPER);

        Input<Void> signal = new Input<>("signal", SignalType.INSTANCE);
        Input<PlayerValue> player = new Input<>("player", PlayerType.INSTANCE);
        Input<Component> message = new Input<>("message", TextType.INSTANCE);

        Output<Void> next = new Output<>("next", SignalType.INSTANCE);

        signal.onSignal((ctx) -> {
            PlayerValue p = player.getValue(ctx);
            if (p.available(ctx)) p.get(ctx).sendMessage(message.getValue(ctx));
            ctx.sendSignal(next);
        });
    }

    @Override
    public Node copy() {
        return new SendMessageNode();
    }
}

File: src/main/kotlin/de/blazemcworld/fireflow/node/impl/player/action/SendMessageNode.kt
package de.blazemcworld.fireflow.node.impl.player.action

import de.blazemcworld.fireflow.node.*
import net.minestom.server.item.Material

object SendMessageNode : BaseNode("Send Message", Material.WRITTEN_BOOK) {
    private val signal = input("Signal", SignalType)
    private val player = input("Player", PlayerType)
    private val message = input("Message", MessageType)
    private val next = output("Next", SignalType)

    override fun setup(ctx: NodeContext) {
        ctx[signal].signalListener = {
            val player = it[ctx[player]]?.resolve()
            val msg = it[ctx[message]]
            if (msg != null && player != null) player.sendMessage(msg)
            it.emit(ctx[next])
        }
    }
}